analyzers/tests/SonarAnalyzer.UnitTest/TestFramework/SymbolicExecution/SymbolicValueAssertions.cs

```diff
@@ -0,0 +1,164 @@
+                .Then
+                .Given(() => Subject.AllConstraints.ToList())
+                .ForCondition(allConstraints => allConstraints.Count != 0)
+                .FailWith("Expected {context:symbolicValue} to have constraint {0}{reason}, but symbolicValue has no constraints.", constraint)
+                .Then
+                .ForCondition(allConstraints => allConstraints.Contains(constraint))
+                .FailWith("Expected {context:symbolicValue} to have constraint {0}{reason}, but symbolicValue has {1} constraints.",
+                    _ => constraint,
+                    x => x);
+            var matchedConstrainted = Subject.AllConstraints.First(x => x == constraint);
+            return new(this, matchedConstrainted);
+        }
+
+        public AndWhichConstraint<SymbolicValueAssertions, SymbolicConstraint> HaveOnlyConstraint(SymbolicConstraint constraint, string because = "", params object[] becauseArgs)
+        {
+            Execute.Assertion
+                .BecauseOf(because, becauseArgs)
+                .ForCondition(Subject != null)
+                .FailWith("The {context:symbolicValue} is not present and can not have constraint {0}.", constraint)
+                .Then
+                .Given(() => Subject.AllConstraints.ToList())
+                .ForCondition(allConstraints => allConstraints.Count > 0)
+                .FailWith("Expected {context:symbolicValue} to have constraint {0}{reason}, but symbolicValue has no constraints.", constraint)
+                .Then
+                .ForCondition(allConstraints => allConstraints.Count == 1)
+                .FailWith("Expected {context:symbolicValue} to have only constraint {0}{reason}, but symbolicValue has {1} constraints.",
+                    _ => constraint,
+                    x => x)
+                .Then
+                .ForCondition(allConstraints => allConstraints[0] == constraint)
+                .FailWith("Expected {context:symbolicValue} to have constraint {0}{reason}, but symbolicValue has {1} constraint.",
+                    _ => constraint,
+                    allConstraints => allConstraints[0]);
+            var matchedConstrainted = Subject.AllConstraints.First(x => x == constraint);
+            return new(this, matchedConstrainted);
+        }
+
+        public AndConstraint<SymbolicValueAssertions> HaveNoConstraints(string because = "", params object[] becauseArgs)
+        {
+            Execute.Assertion
+                .BecauseOf(because, becauseArgs)
+                .ForCondition(Subject == null || Subject.AllConstraints.All(x => x == null))
+                .FailWith("Expected {context:symbolicValue} to have no constraints{reason}, but {0} constraints were found.", Subject?.AllConstraints);
+            return new(this);
+        }
+
+        public AndConstraint<SymbolicValueAssertions> HaveConstraints(params SymbolicConstraint[] constraints)
```

Comment:
Does it require at least these, any of these, or exactly and only these? It's not clear from the name.

What about changing the name to `HaveExactlyConstraints` (or similar) and update the logic not to expect anything additional. To have complete picture of what was added by engine and unexpected by old UTs

Reply:
And now I've found `HaveOnlyConstraints`. So do we really need `HaveConstraints` variant?

Reply:
>  So do we really need HaveConstraints variant?

I can't tell what other constraints we will add in the future (`NumberConstraint.Zero` or `StringConstraint.Empty`). `HaveConstraints` would be fault tolerant to new constraints which is either a good thing or a bad thing depending on your point of view.

Reply:
It's a good thing to fail old UTs if we change constraints. As there might be 
* surprising good consequences
* surprising bad consequences
In both cases, it's good to be explicit about full scope of the UT. 

At the same time, it enforces that the UTs are narrow-focused.